hledger-lib/Hledger/Data/Dates.hs

```diff
@@ -736,31 +854,62 @@ reportinginterval = choice' [
                           return $ Months 2,
                        do string "every"
                           many spacenonewline
-                          n <- fmap read $ some digitChar
-                          thsuffix
+                          n <- nth
                           many spacenonewline
                           string "day"
                           many spacenonewline
-                          string "of"
+                          of_ "week"
+                          return $ DayOfWeek n,
+                       do string "every"
+                          many spacenonewline
+                          n <- weekday
                           many spacenonewline
-                          string "week"
                           return $ DayOfWeek n,
                        do string "every"
                           many spacenonewline
-                          n <- fmap read $ some digitChar
-                          thsuffix
+                          n <- nth
                           many spacenonewline
                           string "day"
-                          optional $ do
-                            many spacenonewline
-                            string "of"
-                            many spacenonewline
-                            string "month"
-                          return $ DayOfMonth n
+                          many spacenonewline
+                          optional $ of_ "month"
+                          return $ DayOfMonth n,
+                       do string "every"
+                          many spacenonewline
+                          d_o_y <- makePermParser $
```

Comment:
It might become more readable with

```haskell
d_o_y <- makePermParser $ DayOfYear <$$> month <||> nth

month = choice' [month, mon] >>= (_, m, _) -> return m
```

Reply:
makes sense, done